community/detectors/xwiki_cve_2024_21650/src/test/java/com/google/tsunami/plugins/detectors/rce/cve202421650/Cve202421650DetectorTest.java

```diff
@@ -0,0 +1,144 @@
+
+  static final String CSRF_TEMPLATE =
+      "<input type=\"hidden\" name=\"form_token\" value=\"XJJ8bxI3PjfwK9FxAUPFCg\" />";
+
+  private MockWebServer mockWebServer;
+  private MockWebServer mockCallbackServer;
+
+  @Inject private Cve202421650Detector detector;
+
+  private NetworkService service;
+  private TargetInfo targetInfo;
+
+  @Bind(lazy = true)
+  @OobSleepDuration
+  private int sleepDuration = 1;
+
+  @Before
+  public void setUp() {
+    mockWebServer = new MockWebServer();
+    mockCallbackServer = new MockWebServer();
+
+    Guice.createInjector(
+            new FakeUtcClockModule(fakeUtcClock),
+            new HttpClientModule.Builder().build(),
+            FakePayloadGeneratorModule.builder().setCallbackServer(mockCallbackServer).build(),
+            new Cve202421650DetectorBootstrapModule())
+        .injectMembers(this);
+
+    targetInfo =
+        TargetInfo.newBuilder()
+            .addNetworkEndpoints(forHostname(mockWebServer.getHostName()))
+            .build();
+    service =
+        NetworkService.newBuilder()
+            .setNetworkEndpoint(
+                forHostnameAndPort(mockWebServer.getHostName(), mockWebServer.getPort()))
+            .setTransportProtocol(TransportProtocol.TCP)
+            .setServiceName("http")
+            .build();
+  }
+
+  @After
+  public void tearDown() throws IOException {
+    mockWebServer.shutdown();
+  }
+
+  @Test
+  public void detect_whenVulnerable_returnsVulnerability() throws IOException {
+    mockWebServer.enqueue(
+        new MockResponse().setResponseCode(HttpStatus.OK.code()).setBody(CSRF_TEMPLATE));
+    mockWebServer.enqueue(
+        new MockResponse().setResponseCode(HttpStatus.OK.code()).setBody(RESPONSE_STRING));
+    mockWebServer.enqueue(new MockResponse().setResponseCode(HttpStatus.ACCEPTED.code()));
+    mockCallbackServer.enqueue(PayloadTestHelper.generateMockSuccessfulCallbackResponse());
+
+    DetectionReportList detectionReports = detector.detect(targetInfo, ImmutableList.of(service));
+
+    assertThat(detectionReports.getDetectionReportsList())
+        .containsExactly(detector.buildDetectionReport(targetInfo, service));
+  }
+
+  @Test
+  public void detect_whenVulnerable_noCallbackServer_returnsVulnerability() throws IOException {
+    mockCallbackServer.shutdown();
+    mockWebServer.enqueue(
+        new MockResponse().setResponseCode(HttpStatus.OK.code()).setBody(CSRF_TEMPLATE));
+    mockWebServer.enqueue(
+        new MockResponse().setResponseCode(HttpStatus.OK.code()).setBody(RESPONSE_STRING));
+    mockWebServer.enqueue(new MockResponse().setResponseCode(HttpStatus.ACCEPTED.code()));
+    mockCallbackServer.enqueue(PayloadTestHelper.generateMockSuccessfulCallbackResponse());
+
+    DetectionReportList detectionReports = detector.detect(targetInfo, ImmutableList.of(service));
+
+    assertThat(detectionReports.getDetectionReportsList())
+        .containsExactly(detector.buildDetectionReport(targetInfo, service));
+  }
```

Comment:
Can you check the correctness of this test? It looks wrong to me, as by reading the code, it looks like when the callback server is not available, no findings should be returned. I think the issue here is stopping the callback server, as it seems it causes an exception that is not handled.

I'm running the tests with `./gradlew test --info --rerun-tasks`

Reply:
Ok, I found some problems around callback server usage. I fixed it through detector optimisation & tests improvement. Big changes there. I use spying of random values there, so I did revert mockito dependancy.
Now tests must work as expected.